### ISSUE TITLE
SystemJS doesn't load RxJS properly

### ISSUE THREAD

**Issue by dceleda** [systemjs/systemjs#1675]
https://stackoverflow.com/questions/43462157/systemjs-doesnt-load-rxjs-properly/43789727#43789727

**Comment by guybedford** [comment]
Did you try `import Observable from 'rxjs/Observable';`?

**Comment by dceleda** [comment]
Yes, I've tried many different ways of importing.

It helped when I changed to older version of systemjs (0.19.27).

**Comment by guybedford** [comment]
This sounds like it could be related to the 0.20.11 bug. Perhaps try SystemJS 0.20.10 and see if that works here.

I am considering some alternatives here for a future release as this does seem a common catch.

**Comment by EricSimons** [comment]
@guybedford this is still an issue on 0.20.10. I've noticed this problem with a handful of CJS based imports — System will put all named exports _underneath_ the default export. So your example above would yield:

```
import Observable from 'rxjs/Observable';
console.log(Observable);
// > Object {Observable: function}
```

Whereas you'd want the function itself to be returned, not inside of an object with the identical name.

I also can confirm that this works properly in 0.19.x.

Any idea when a fix for this will land for 0.20.x? (i.e. should I roll back my current app to 0.19.x for the foreseeable future?)

**Comment by guybedford** [comment]
Did you try SystemJS 0.20.13 here? I believe I added a fix for this, let me
know if it works for you.

**Comment by EricSimons** [comment]
@guybedford yup, I'm currently working off of 0.20.13 and it doesn't work. I think your fix in 0.20.13 was for proper handling of default exports, right? Whereas this seems to be an issue of _non_-default exports.

**Comment by EricSimons** [comment]
To repro, here's a sample config file:

```json
{  
   "paths":{  
      "npm:":"https://unpkg.com/",
   },
   "map":{  
      "rxjs":"npm:rxjs@5.3.1",
      "symbol-observable@1.0.4":"npm:symbol-observable@1.0.4",
   },
   "packages":{  
      "rxjs":{  
         "defaultExtension":"js",
         "main":"Rx.js",
         "map":{  
            "symbol-observable":"npm:symbol-observable@1.0.4"
         }
      }
   }
}
```

Then try calling System.import on a file that contains:

```js
import Observable from 'rxjs/Observable';
console.log(Observable);
```

And you'll get: `> Object {Observable: function}` instead of `> function`

**Comment by EricSimons** [comment]
@guybedford were you able to repro this on your end? Happy to set up a live example for you if need be

**Comment by EricSimons** [comment]
@guybedford I think I found the (general) location of the problem. At https://github.com/ModuleLoader/es-module-loader/blob/ad8ca7dc61a2edffa12972b47bffb31da5a9e49f/core/register-loader.js#L640:

```
if (moduleDefault && moduleDefault.__esModule) {
  for (var p in moduleDefault) {
    if (Object.hasOwnProperty.call(moduleDefault, p))
      moduleObj[p] = moduleDefault[p];
  }
}
```

I think the problem is that many CJS modules don't specify __esModule, but the user would still expect to receive named exports. I changed it locally to not look for _esModule and check if the moduleObject already has a named export before assigning the exports from default over to it:

```
if (moduleDefault) {
  for (var p in moduleDefault) {
    if (Object.hasOwnProperty.call(moduleDefault, p) && !Object.hasOwnProperty.call(moduleObj, p))
      moduleObj[p] = moduleDefault[p];
    }
 }
```

This is obviously a janky solution, but I've got it working locally now 👍

**Comment by guybedford** [comment]
@EricSimons we don't do this automatically because the ecosystem has to behave the same way as NodeJS will when ES modules are supported there, and NodeJS won't be doing this auto named exports extension.

If you want to configure this use meta:

```js
System.config({
  meta: {
    'path/to/rxjs/*': { esModule: true }
  }
});
```

**Comment by EricSimons** [comment]
@guybedford I added the meta but same result. We're just going to use a forked version of system for the time being — thanks for the help

**Comment by jamesbrobb** [comment]
@guybedford I've just updated from `0.19.41` to `0.20.17` and am now experiencing this issue as well. As @EricSimons mentioned, the suggested solution of adding `{ esModule: true }` unfortunately doesn't fix is it. Is there any update?

**Comment by nachtgold** [comment]
I found a similar problem and filed it here https://github.com/ReactiveX/rxjs/issues/2815. Maybe it helps fixing the bug. I used rxjs@5.4.3 loaded by systemjs@0.20.18 and compiled by typescript 2.4.2 and got the error `ERROR TypeError: Subject_1.Subject is not a constructor`

**Comment by inquiztr** [comment]
I was following this thread as I also had the Subject is not a constructor issue after upgrading to the latest JSPM/SystemJS 0.20.18 but I did find adding the Meta as @guybedford suggested, did work for me.

**Comment by inquiztr** [comment]
Ok one issue.. While adding the meta - esmodule : true works when unbundled, the error comes back once i bundle everything up for production.  I am using : 
meta: {
        "/jspm_packages/npm/rxjs@5.2.0/*": {
            "esModule": true
        }
    },

Is there a better way to do this.  A) a way that this setting would work when even bundled b) a way to specify the path without the version number hard coded in?

**Comment by nachtgold** [comment]
What do you do/use for "production bundling"?

**Comment by inquiztr** [comment]
I am using JSPM 0.17.0-beta.44 which uses System Builder 0.20.18.  Ended up defining a 'package' for rxjs 
```
"/jspm_packages/npm/rxjs@5.2.0": {
            "main": "Rx.js",
            "defaultExtension": "js",
            "format": "cjs",
            "meta": {
                "./*": {
                    "esModule": true
                }
            }
        }
```
Which is also working unbundled, but when bundled using System Builder, esModule isn't working. I found a bug referencing this here:  [https://github.com/systemjs/builder/issues/775](url) and was fixed in System Builder 0.16.5. Unfortunately the latest JSPM beta is still using System Builder 0.16.10

So now I need to figure out how to override jspms dependancy and use the latest SB. I guess I will have to take my question over to the JSPM repo.

**Comment by kylecordes** [comment]
I came across this issue are researching the following problem: As far as I can tell, having tried every variation of every setting mentioned the documentation and in a dozen issues, there is no way to set the esModule flag for code which is bundled inside a System bundle. At the same time, the "bundle" loading feature, critical to keep roundtrips under control, only works with System bundles. This appears to be an insurmountable problem for loading RxJS using 0.20/0.21.  @EricSimons mentioned above that he got around it by modifying a couple of lines of System.

(For the first few decades of my software development career, I tried very hard to no hackery, to bring order and consistency and intentionality to all code I touched. Have I given up? Not exactly, but now I could perhaps be persuaded that it really is all hacks, all the way down...)

The following script creates a RxJS System bundle with the esModule flag set inside, so that it be can be consumed by TypeScript code.

```
const fs = require('fs');
const Builder = require("systemjs-builder");
const builder = new Builder('./');

builder.config({
  paths: {
    "rxjs/*": "rxjs/*.js"
  },
  packages: {
    "rxjs": {
      main: 'Rx.js',
      defaultExtension: "js"
    }
  },
  baseURL: "node_modules"
});

builder.bundle('rxjs', { sourceMaps: false }).then(output => {
  const code = output.source
    .split('"use strict";')
    .join('"use strict"; exports.__esModule = true;');
  fs.writeFileSync('lib/rxjs.system.js', code);
});
```
